parsons/zoom/zoom.py

```diff
@@ -275,3 +295,197 @@ def get_webinar_registrants(self, webinar_id):
+            webinar_id: str
+                Unique identifier for Zoom webinar
+            poll_id: int
+                Unique identifier for poll
+
+        `Returns`:
+            Parsons Table of all polling responses
+        """
+
+        endpoint = f"webinars/{webinar_id}/polls/{poll_id}"
+        tbl = self._get_request(endpoint=endpoint, data_key="questions")
+
+        if type(tbl) == dict:
+            logger.debug(f"No poll data returned for poll ID {poll_id}")
+            return Table(tbl)
+
+        logger.info(
+            f"Retrieved {tbl.num_rows} rows of metadata [meeting={webinar_id} poll={poll_id}]"
+        )
+
+        return self.__handle_nested_json(table=tbl, column="prompts")
+
+    def get_webinar_all_polls_metadata(self, webinar_id) -> Table:
+        """
+        Get metadata for all polls for a given webinar ID
+
+        Required scopes: `webinar:read`
+
+        `Args`:
+            webinar_id: str
+                Unique identifier for Zoom webinar
+
+        `Returns`:
+            Parsons Table of all polling responses
+        """
+
+        endpoint = f"webinars/{webinar_id}/polls"
+        tbl = self._get_request(endpoint=endpoint, data_key="polls")
+
+        if type(tbl) == dict:
+            logger.debug(f"No poll data returned for webinar ID {webinar_id}")
+            return Table(tbl)
+
+        logger.info(f"Retrieved {tbl.num_rows} polls for meeting ID {webinar_id}")
+
+        return self.__handle_nested_json(table=tbl, column="questions")
+
+    def get_past_webinar_poll_metadata(self, webinar_id) -> Table:
+        """
+        Retrieves the metadata for Webinar Polls of a specific Webinar
+
+        Required scopes: `webinar:read`
+
+        `Args`:
+            webinar_id: str
+                The webinar's ID or universally unique ID (UUID).
+
+        `Returns`:
+            Parsons Table of all polling responses
+        """
+
+        endpoint = f"past_webinars/{webinar_id}/polls"
+        tbl = self._get_request(endpoint=endpoint, data_key="questions")
+
+        if type(tbl) == dict:
+            logger.debug(f"No poll data returned for webinar ID {webinar_id}")
+            return Table(tbl)
+
+        logger.info(f"Retrieved {tbl.num_rows} polls for meeting ID {webinar_id}")
+
+        return self.__handle_nested_json(table=tbl, column="prompts")
+
+    def get_meeting_poll_results(self, meeting_id) -> Table:
+        """
+        Get a report of poll results for a past meeting
+
+        Required scopes: `report:read:admin`
+        """
+
+        endpoint = f"report/meetings/{meeting_id}/polls"
+        tbl = self._get_request(endpoint=endpoint, data_key="questions")
+
+        if type(tbl) == dict:
+            logger.debug(f"No poll data returned for meeting ID {meeting_id}")
+            return Table(tbl)
+
+        logger.info(f"Retrieved {tbl.num_rows} reults for meeting ID {meeting_id}")
+
+        return self.__handle_nested_json(table=tbl, column="question_details")
```

Comment:
This was better, but it was still putting JSON data into the table for me. I took the liberty of massaging it around until I got a schema that I liked for the data coming back. Here is the code I wound up with:

```
    def __process_poll_results(self, tbl: Table) -> Table:
        tbl.add_column("poll_taker_id", lambda _: str(uuid.uuid4()))
        tbl = tbl.unpack_nested_columns_as_rows(
            "question_details", key="poll_taker_id", expand_original=True
        )
        tbl.remove_column("poll_taker_id")
        tbl.remove_column("question_details")

        tbl = tbl.unpack_dict(
            "question_details_value", include_original=True, prepend=False
        )
        tbl.remove_column("question_details_value")

        return tbl


    def get_meeting_poll_results(self, meeting_id) -> Table:
        """
        Get a report of poll results for a past meeting

        Required scopes: `report:read:admin`
        """

        endpoint = f"report/meetings/{meeting_id}/polls"
        tbl = self._get_request(endpoint=endpoint, data_key="questions")

        if type(tbl) == dict:
            logger.debug(f"No poll data returned for meeting ID {meeting_id}")
            return Table(tbl)

        logger.info(f"Retrieved {tbl.num_rows} reults for meeting ID {meeting_id}")

        if tbl.num_rows > 0:
            tbl = self.__process_poll_results(tbl)

        return tbl

    def get_webinar_poll_results(self, webinar_id) -> Table:
        """
        Get a report of poll results for a past webinar

        Required scopes: `report:read:admin`
        """

        endpoint = f"report/webinars/{webinar_id}/polls"
        tbl = self._get_request(endpoint=endpoint, data_key="questions")

        if type(tbl) == dict:
            logger.debug(f"No poll data returned for webinar ID {webinar_id}")
            return Table(tbl)

        logger.info(f"Retrieved {tbl.num_rows} reults for webinar ID {webinar_id}")

        if tbl.num_rows > 0:
            tbl = self.__process_poll_results(tbl)

        return tbl
```

The data comes out pretty nice, with each answer getting a full separate row to make it easy to query in Python or in SQL

Reply:
Here is what the table in my script looked like after pushing to Redshift (I appended the three columns with the event information.)

![image](https://github.com/move-coop/parsons/assets/3255423/0201d078-4ae0-4389-af25-969620b4263b)